cudax/include/cuda/experimental/__execution/stream/context.cuh

```diff
@@ -0,0 +1,215 @@
+}
+
+template <class _Fn, class... _Args>
+__global__ static void __stream_invoke_r(_CUDA_VSTD::__call_result_t<_Fn, _Args...>* __return, _Fn __fn, _Args... __args)
+{
+  using _Return = _CUDA_VSTD::__call_result_t<_Fn, _Args...>;
+  ::new (static_cast<void*>(__return)) _Return(static_cast<_Fn&&>(__fn)(static_cast<_Args&&>(__args)...));
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////
+// stream_context
+struct _CCCL_TYPE_VISIBILITY_DEFAULT stream_context : private __immovable
+{
+  stream_context() noexcept = default;
+
+  _CCCL_HOST_API void sync() noexcept
+  {
+    __stream_.sync();
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////////////
+  // stream scheduler
+  struct _CCCL_TYPE_VISIBILITY_DEFAULT scheduler
+  {
+    using scheduler_concept = scheduler_t;
+
+    _CCCL_API bool operator==(const scheduler& __other) const noexcept
+    {
+      return __stream_ref_ == __other.__stream_ref_;
+    }
+
+    _CCCL_API static constexpr auto query(get_forward_progress_guarantee_t) noexcept
+    {
+      return forward_progress_guarantee::weakly_parallel;
+    }
+
+    _CCCL_API static constexpr auto query(get_domain_t) noexcept
+    {
+      return stream_domain{};
+    }
+
+    _CCCL_API constexpr auto query(get_stream_t) const noexcept
+    {
+      return __stream_ref_;
+    }
+
+    _CCCL_API auto schedule() const noexcept
+    {
+      return __sndr_t{{}, {__stream_ref_}};
+    }
+
+    stream_ref __stream_ref_;
+  };
+
+  _CCCL_TRIVIAL_HOST_API auto get_scheduler() noexcept -> scheduler
+  {
+    return scheduler{__stream_};
+  }
+
+  _CUDAX_SEMI_PRIVATE :
+  ////////////////////////////////////////////////////////////////////////////////////////
+  // environment of the stream scheduler's sender
+  struct _CCCL_TYPE_VISIBILITY_DEFAULT __env_t
+  {
+    _CCCL_API constexpr auto query(get_completion_scheduler_t<set_value_t>) const noexcept
+    {
+      return scheduler{__stream_ref_};
+    }
+
+    _CCCL_API constexpr auto query(get_stream_t) const noexcept
+    {
+      return __stream_ref_;
+    }
+
+    _CCCL_TRIVIAL_HOST_API static constexpr auto query(get_domain_t) noexcept
+    {
+      return stream_domain{};
+    }
+
+    stream_ref __stream_ref_;
+  };
+
+  ////////////////////////////////////////////////////////////////////////////////////////
+  // stream scheduler's operation state
+  template <class _Rcvr>
+  struct _CCCL_TYPE_VISIBILITY_DEFAULT __opstate_t
+  {
+    using operation_state_concept = operation_state_t;
+
+    _CCCL_API __opstate_t(_Rcvr __rcvr, stream_ref __stream_ref) noexcept
+        : __rcvr_{static_cast<_Rcvr&&>(__rcvr)}
+        , __stream_ref_{__stream_ref}
+    {}
+
+    _CCCL_IMMOVABLE_OPSTATE(__opstate_t);
+
+    _CCCL_API void start() & noexcept
+    {
+      NV_IF_TARGET(NV_IS_DEVICE, (__device_start();), (__host_start();));
+    }
+
+  private:
+    _CCCL_HOST_API void __host_start() noexcept
```

Comment:
No change requested as we currently do not support cudax with nvrtc or non-cuda compiler, but this should be guarded by `_CCCL_HOST_COMPILETION()` and the device function by `_CCCL_DEVICE_COMPILATION()`

Reply:
why is that?

Reply:
Because this would break NVRTC due to it being a device function. The other definition would probably break any host only compiler because it relies on cuda properties

Reply:
wdym nvrtc can't compile a device function?! as for the host function, yes a c++ compiler doesn't understand `<<<>>>` syntax, but this file can only be compiled as CUDA. it's a CUDA stream scheduler. `nvcc` makes it so that the host compiler never sees the `<<<>>>`.